Corona-Warn-App/src/main/java/de/rki/coronawarnapp/bugreporting/censors/BugCensor.kt

```diff
@@ -1,47 +1,85 @@
 package de.rki.coronawarnapp.bugreporting.censors
 
-import kotlin.math.max
-import kotlin.math.min
-
 interface BugCensor {
 
     /**
      * If there is something to censor a new log line is returned, otherwise returns null
      */
-    suspend fun checkLog(message: String): CensoredString?
+    suspend fun checkLog(message: String): CensorContainer?
 
-    data class CensoredString(
-        // The censored version of the string
-        val string: String,
-        // The range that we censored
-        // If there is a collision, this range in the original needs to be removed.
-        val range: IntRange? = null
-    )
+    data class CensorContainer(
+        // Original String, necessary for correct censoring ranges
+        val original: String,
+        val actions: List<Action> = emptyList()
+    ) {
 
-    companion object {
-        operator fun CensoredString.plus(newer: CensoredString?): CensoredString {
-            if (newer == null) return this
+        fun censor(toReplace: String, replacement: String): CensorContainer {
+            if (!original.contains(toReplace)) return this
 
-            val range = when {
-                newer.range == null -> this.range
-                this.range == null -> newer.range
-                else -> min(this.range.first, newer.range.first)..max(this.range.last, newer.range.last)
+            val start = original.indexOf(toReplace)
+            if (start == -1) return this // Shouldn't happen
+
+            val end = original.lastIndexOf(toReplace) + toReplace.length
+
+            val newAction = Action(
+                range = start..end,
+                execute = { it.replace(toReplace, replacement) }
+            )
+            return this.copy(actions = actions.plus(newAction))
+        }
+
+        fun compile(): CensoredString? {
+            val ranges = actions.map { it.range }
+            if (ranges.isEmpty()) return null
+
+            val isIntersecting = ranges.any { outter ->
+                ranges.any { inner ->
+                    outter != inner &&
+                        (inner.contains(outter.first) || inner.contains(outter.last)) &&
+                        (inner.last != outter.first && inner.first != outter.last)
+                }
             }
 
-            return CensoredString(string = newer.string, range = range)
+            return if (isIntersecting) {
+                val minMin = ranges.minOf { it.first }.coerceAtLeast(0).coerceAtMost(original.length)
```

Comment:
I think we don't need `coerce` here, since we calculate the range bounds of the `original` LogLine here and so we should never get an `IndexOutOfBounds` exception.  🤔